pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,527 @@
+				glog.V(1).Infof("EndpointsTracker: Setting endpoints for %s to %+v", svcPort, newEndpoints)
+				state.endpoints = newEndpoints
+			}
+			registeredEndpoints[svcPort] = true
+		}
+	}
+	// Remove endpoints missing from the update.
+	for service := range proxier.serviceMap {
+		if _, exists := registeredEndpoints[service]; !exists {
+			glog.V(2).Infof("EndpointsTracker: Removing endpoints for %s", service)
+			proxier.serviceMap[service].endpoints = []string{}
+		}
+	}
+}
+
+// used in OnEndpointsUpdate
+type hostPortPair struct {
+	host string
+	port int
+}
+
+func isValidEndpoint(hpp *hostPortPair) bool {
+	return hpp.host != "" && hpp.port > 0
+}
+
+// Tests whether two slices are equivalent.  This sorts both slices in-place.
+func slicesEquiv(lhs, rhs []string) bool {
+	if len(lhs) != len(rhs) {
+		return false
+	}
+	if reflect.DeepEqual(slice.SortStrings(lhs), slice.SortStrings(rhs)) {
+		return true
+	}
+	return false
+}
+
+func flattenValidEndpoints(endpoints []hostPortPair) []string {
+	// Convert Endpoint objects into strings for easier use later.  Ignore
+	// the protocol field - we'll get that from the Service objects.
+	var result []string
+	for i := range endpoints {
+		hpp := &endpoints[i]
+		if isValidEndpoint(hpp) {
+			result = append(result, net.JoinHostPort(hpp.host, strconv.Itoa(hpp.port)))
+		}
+	}
+	return result
+}
+
+// servicePortToServiceChain takes the ServicePortName for a
+// service and returns the associated iptables chain
+// this is computed by hashing (sha256) then encoding to base64 and
+// truncating with the prefix "KUBE-SVC-"
+// We do this because Iptables Chain Names must be <= 28 chars long
+func servicePortToServiceChain(s proxy.ServicePortName) utiliptables.Chain {
+	hash := sha256.Sum256([]byte(s.String()))
+	encoded := base64.URLEncoding.EncodeToString(hash[:])
+	return utiliptables.Chain("KUBE-SVC-" + encoded[:19])
+}
+
+// this is the same as servicePortToServiceChain but with the endpoint included essentially
+func servicePortAndEndpointToServiceChain(s proxy.ServicePortName, endpoint string) utiliptables.Chain {
+	hash := sha256.Sum256([]byte(s.String() + "_" + endpoint))
+	encoded := base64.URLEncoding.EncodeToString(hash[:])
+	return utiliptables.Chain("KUBE-SVC-" + encoded[:19])
+}
+
+// This is where all of the iptables-save/restore calls happen.
+// The only other iptables rules are those that are setup in iptablesInit()
+// assumes proxier.mu is held
+func (proxier *Proxier) syncProxyRules() error {
+	glog.V(2).Infof("Syncing iptables rules.")
+
+	// ensure main chain and rule connecting to output
+	args := []string{"-m", "comment", "--comment", "handle services"}
+	if _, err := proxier.iptables.EnsureChain(utiliptables.TableNAT, iptablesServicesChain); err != nil {
+		return err
+	}
+	if _, err := proxier.iptables.EnsureRule(utiliptables.Prepend, utiliptables.TableNAT, utiliptables.ChainOutput, append(args, "-j", string(iptablesServicesChain))...); err != nil {
```

Comment:
Ah, yeah we probably do, I was just following the pseudoish-code from the
redesign you suggested here:
https://github.com/GoogleCloudPlatform/kubernetes/pull/9210#issuecomment-128912554
I'll fix that next commit.

On Tue, Aug 11, 2015 at 12:44 AM, Tim Hockin <notifications@github.com>
wrote:

> In pkg/proxy/iptables/proxier.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210#discussion_r36711389>
> :
>
> > +	encoded := base64.URLEncoding.EncodeToString(hash[:])
> > +	return utiliptables.Chain("KUBE-SVC-" + encoded[:19])
> > +}
> > +
> > +// This is where all of the iptables-save/restore calls happen.
> > +// The only other iptables rules are those that are setup in iptablesInit()
> > +// assumes proxier.mu is held
> > +func (proxier *Proxier) syncProxyRules() error {
> > +	glog.V(2).Infof("Syncing iptables rules.")
> > +
> > +	// ensure main chain and rule connecting to output
> > +	args := []string{"-m", "comment", "--comment", "handle services"}
> > +	if _, err := proxier.iptables.EnsureChain(utiliptables.TableNAT, iptablesServicesChain); err != nil {
> > +		return err
> > +	}
> > +	if _, err := proxier.iptables.EnsureRule(utiliptables.Prepend, utiliptables.TableNAT, utiliptables.ChainOutput, append(args, "-j", string(iptablesServicesChain))...); err != nil {
>
> you need to do this for OUTPUT and PREROUTING don't you? How does traffic
> from containers get handled otherwise?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210/files#r36711389>
> .
>